### PR TITLE
[GOBBLIN-2110]Made retry_exception_predicate configurable in RetryerFactory

### PR DESCRIPTION
**Pull request by pratapaditya04**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-2110

### Description
- [x] Here are some details about my PR, including screenshots (if applicable):
RetryerFactory does not give option to set retry_exception_predicate.Made it configurable

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-utility/src/main/java/org/apache/gobblin/util/retry/RetryerFactory.java [hunk 1]
```diff
@@ -53,10 +59,14 @@ public class RetryerFactory<T> {
   // value large or equal to 1
   public static final String RETRY_TIMES = "retry_times";
 
-  private static final Predicate<Throwable> RETRY_EXCEPTION_PREDICATE;
+  // Key to store a comma-separated list of exception class names that should be retried
+  public static final String EXCEPTION_LIST_FOR_RETRY_CONFIG_KEY = "EXCEPTION_LIST_FOR_RETRY";
```
Comment (arjun4084346) [inline gobblin-utility/src/main/java/org/apache/gobblin/util/retry/RetryerFactory.java:63]: maybe we can move this line along with https://github.com/apache/gobblin/blob/master/gobblin-api/src/main/java/org/apache/gobblin/configuration/ConfigurationKeys.java#L1063

File: gobblin-utility/src/main/java/org/apache/gobblin/util/retry/RetryerFactory.java [hunk 2]
```diff
@@ -87,15 +97,18 @@ public static <T> Retryer<T> newInstance(Config config, Optional<RetryListener>
...
+    Predicate<Throwable> retryPredicate = getRetryPredicateFromConfigOrDefault(config);
+
     switch (type) {
       case EXPONENTIAL:
-        builder = newExponentialRetryerBuilder(config);
+        builder = newExponentialRetryerBuilder(config, retryPredicate);
         break;
       case FIXED:
-        builder = newFixedRetryerBuilder(config);
+        builder = newFixedRetryerBuilder(config, retryPredicate);
```
Comment (arjun4084346) [inline gobblin-utility/src/main/java/org/apache/gobblin/util/retry/RetryerFactory.java:108]: lets not add something to the param that can be constructer using the existing param.
lets move `Predicate<Throwable> retryPredicate = getRetryPredicateFromConfigOrDefault(config);` inside `newFixedRetryerBuilder` and other builders

File: gobblin-utility/src/main/java/org/apache/gobblin/util/retry/RetryerFactory.java [hunk 3]
```diff
@@ -104,32 +117,75 @@ public static <T> Retryer<T> newInstance(Config config, Optional<RetryListener>
...
+   @param config the configuration object containing the list of exception class names
+   @return a Predicate that evaluates to true if the throwable should be retried, false otherwise
+   */
+  @VisibleForTesting
+  public static Predicate<Throwable> getRetryPredicateFromConfigOrDefault(Config config) {
+    // Retrieve the list of exception class names from the configuration
+    List<String> exceptionList = ConfigUtils.getStringList(config, EXCEPTION_LIST_FOR_RETRY_CONFIG_KEY);
+
+    // If the exception list is null or empty, return the default retry predicate
+    if (exceptionList == null || exceptionList.isEmpty()) {
```
Comment (arjun4084346) [inline gobblin-utility/src/main/java/org/apache/gobblin/util/retry/RetryerFactory.java:133]: can use `Strings.isNullOrEmpty()`

Comment (arjun4084346) [reply gobblin-utility/src/main/java/org/apache/gobblin/util/retry/RetryerFactory.java:133]: in this case, actually lets create a method `List<String> getStringList(Config config, String path, List<String> def)` (taking idea from the similar methods to get Int, String or other types) and then

`return ConfigUtils.getStringList(config, EXCEPTION_LIST_FOR_RETRY_CONFIG_KEY, RETRY_EXCEPTION_PREDICATE_DEFAULT);`

File: gobblin-utility/src/main/java/org/apache/gobblin/util/retry/RetryerFactory.java [hunk 4]
```diff
@@ -104,32 +117,75 @@ public static <T> Retryer<T> newInstance(Config config, Optional<RetryListener>
...
+          try {
+            Class<?> clazz = Class.forName(exceptionClassName);
+            if (Exception.class.isAssignableFrom(clazz)) {
+              // Return a Predicate that checks if a Throwable is an instance of the class
+              return (Predicate<Throwable>) clazz::isInstance;
+            } else {
+              LOG.error("{} is not an exception.", exceptionClassName);
+            }
+          } catch (ClassNotFoundException exception) {
+            LOG.error("Class not found for the given exception className {}", exceptionClassName, exception);
```
Comment (arjun4084346) [inline gobblin-utility/src/main/java/org/apache/gobblin/util/retry/RetryerFactory.java:148]: LOG.error("Class not found for the given exception className {}. Ignoring it.", exceptionClassName, exception);

File: gobblin-utility/src/main/java/org/apache/gobblin/util/retry/RetryerFactory.java [hunk 5]
```diff
@@ -104,32 +117,75 @@ public static <T> Retryer<T> newInstance(Config config, Optional<RetryListener>
...
+          } catch (Exception exception) {
+            LOG.error("Failed to instantiate exception {}", exceptionClassName, exception);
```
Comment (arjun4084346) [inline gobblin-utility/src/main/java/org/apache/gobblin/util/retry/RetryerFactory.java:150]: same here, "Ignoring"

File: gobblin-utility/src/main/java/org/apache/gobblin/util/retry/RetryerFactory.java [hunk 6]
```diff
@@ -104,32 +117,75 @@ public static <T> Retryer<T> newInstance(Config config, Optional<RetryListener>
...
+          }
+          return null;
+        }).filter(Objects::nonNull) // Filter out any null values
+        .reduce(com.google.common.base.Predicates::or) // Combine all predicates using logical OR
+        .orElse(RETRY_EXCEPTION_PREDICATE_DEFAULT); // Default to retryExceptionPredicate if no valid predicates are found
+
+    return retryPredicate;
```
Comment (arjun4084346) [inline gobblin-utility/src/main/java/org/apache/gobblin/util/retry/RetryerFactory.java:157]: you can return retryPredicate inline

File: gobblin-utility/src/test/java/org/apache/gobblin/util/retry/RetryerFactoryTest.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+    Predicate<Throwable> result = RetryerFactory.getRetryPredicateFromConfigOrDefault(config);
+
+    Assert.assertEquals(RetryerFactory.RETRY_EXCEPTION_PREDICATE_DEFAULT, result);
+  }
+
+  @Test
+  public void testGetRetryPredicateFromConfigOrDefault_withMixedExceptions() {
+    Config config = Mockito.mock(Config.class);
+    Mockito.when(config.hasPath(EXCEPTION_LIST_FOR_RETRY_CONFIG_KEY)).thenReturn(true);
```
Comment (arjun4084346) [inline gobblin-utility/src/test/java/org/apache/gobblin/util/retry/RetryerFactoryTest.java:71]: you can create Config like `ConfigBuilder configBuilder = ConfigBuilder.create()
        .addPrimitive(ConfigurationKeys.FLOW_GROUP_KEY, flowConfig.getId().getFlowGroup())
        .addPrimitive(ConfigurationKeys.FLOW_NAME_KEY, flowConfig.getId().getFlowName());`
        
        or like `Config config = ConfigFactory.empty().withValue(GobblinClusterConfigurationKeys.GOBBLIN_CLUSTER_SYSTEM_PROPERTY_PREFIX + ".prop2",
        ConfigValueFactory.fromAnyRef("val2"))`

File: gobblin-utility/src/main/java/org/apache/gobblin/util/retry/RetryerFactory.java (hunk 1 above)
Comment (pratapaditya04) [reply gobblin-utility/src/main/java/org/apache/gobblin/util/retry/RetryerFactory.java]: makes sense, addressed

File: gobblin-utility/src/main/java/org/apache/gobblin/util/retry/RetryerFactory.java (hunk 2 above)
Comment (pratapaditya04) [reply gobblin-utility/src/main/java/org/apache/gobblin/util/retry/RetryerFactory.java]: addressed

File: gobblin-utility/src/main/java/org/apache/gobblin/util/retry/RetryerFactory.java (hunk 3 above)
Comment (pratapaditya04) [reply gobblin-utility/src/main/java/org/apache/gobblin/util/retry/RetryerFactory.java]: Used CollectionUtils.IsEmpty()

File: gobblin-utility/src/test/java/org/apache/gobblin/util/retry/RetryerFactoryTest.java (hunk 7 above)
Comment (pratapaditya04) [reply gobblin-utility/src/test/java/org/apache/gobblin/util/retry/RetryerFactoryTest.java]: addressed

File: gobblin-api/src/main/java/org/apache/gobblin/configuration/ConfigurationKeys.java [hunk 8]
```diff
@@ -1061,6 +1061,8 @@ public class ConfigurationKeys {
   // describes a comma separated list of non transient errors that may come in a gobblin job
   // e.g. "invalid_grant,CredentialStoreException"
   public static final String GOBBLIN_NON_TRANSIENT_ERRORS = "gobblin.errorMessages.nonTransientErrors";
+  // Key to store a comma-separated list of exception class names that should be retried
+  public static final String EXCEPTION_LIST_FOR_RETRY_CONFIG_KEY = "EXCEPTION_LIST_FOR_RETRY";
```
Comment (arjun4084346) [inline gobblin-api/src/main/java/org/apache/gobblin/configuration/ConfigurationKeys.java:1065]: approving for now, but lets change the RHS to small case in the next commit and future prs

**Review by arjun4084346 — APPROVED** [review]

**Pull request merged**